libp2p/switch.nim

```diff
@@ -247,19 +251,13 @@ proc start*(s: Switch) {.async, gcsafe.} =
     )
 
     if addrs.len > 0:
-      startFuts.add(t.start(addrs))
-
-  await allFutures(startFuts)
-
-  for s in startFuts:
-    if s.failed:
-      # TODO: replace this exception with a `listenError` callback. See
-      # https://github.com/status-im/nim-libp2p/pull/662 for more info.
-      raise newException(transport.TransportError,
-        "Failed to start one transport", s.error)
+      try:
+        await t.start(addrs)
```

Comment:
In the spirit of 661, I would give the callback as parameter to the start, and let the transport handle the errors

Btw, I completely missed that, but right now if an address fail, we'll not clean up properly (if a transport is already started, it will not be stopped). So we need to stop the switch on error

Reply:
We need to pass the `listenError` to the transport, otherwise there is no way of handling this at the correct granularity. Transports can bind many addresses and any one could fail, so it only makes sense to handle it in the transport. 

Maybe, we should even consider setting the callback on the transport itself and allow each transport to use a different callback, then each transport's `start()` would throw or continue depending on what `listenError` returns.

Reply:
Hmm, thats a good point, but I think it's orthogonal to this particular issue. 

I would reframe it as - should we treat unhandled addresses as a failure case or not? Having an address that isn't handled by any one transport doesn't seem like a candidate for a failure, but we should definitely warn the user about it.

On the other hand, nothing prevents us from passing a callback to both the transports and the switch, which could be another way of allowing the consumer to handle all potential failure/exceptional scenarios.

Reply:
> should we treat unhandled addresses as a failure case or not

In the latest commit, I removed addresses from tcptransport that were not handled in start (https://github.com/status-im/nim-libp2p/blob/cb4298afcaf2e225b0b8df5483f636ebf12aa3a1/libp2p/transports/tcptransport.nim#L151). The idea being that after the switch calls `start`, it could compare which addresses are remaining in the transport to know the addresses that have not been handled, then treat as an error if needed. Alternatively, we could filter out unhandlable addresses in the transport constructor, and then the switch could do the comparison at any point (not just after `start`)

Reply:
> We need to pass the listenError to the transport

The latest commit now has support for this. Check out [the tests](https://github.com/status-im/nim-libp2p/pull/668/files#diff-4ce97f016e8c6ea2e9574d1355ff0a2c3657e78e3ee971556ebce4b1f5a94bdeR1040-R1180) for implementation. The two tests only cover fully pessimistic or fully optimistic (where both `switch.listenError` and `transport.listenError` are either pessimistic or optimistic), however it's possible to have partially pessimistic/optimistic where `switch.listenError` is pessimistic and `transport.listenError` is optimistic and vice versa. I just haven't had a chance to add those tests yet.

Reply:
Another option would be to simply gives the callback to start, like:

```nim
proc start*(s: Switch, errorHandler = ListenErrorDefault) {.async, gcsafe.} =
[...]
  await t.start(addrs, errorHandler)

[tcptransport.nim]
method start*(
  self: TcpTransport,
  addrs: seq[MultiAddress],
  errorHandler: ListenErrorCallback) {.async.} =
```
wdyt?

Not sure whats more convenient as a user